src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -1,81 +1,131 @@
+import io.iohk.ethereum.domain.BlockchainWriter
+import io.iohk.ethereum.domain.ChainWeight
+import io.iohk.ethereum.ledger.BlockData
+import io.iohk.ethereum.ledger.BlockExecution
+import io.iohk.ethereum.ledger.BlockExecutionError
 import io.iohk.ethereum.ledger.BlockExecutionError.MPTError
 import io.iohk.ethereum.ledger.BlockExecutionError.ValidationBeforeExecError
+import io.iohk.ethereum.ledger.BlockExecutionSuccess
+import io.iohk.ethereum.ledger.BlockMetrics
+import io.iohk.ethereum.ledger.BlockQueue
 import io.iohk.ethereum.ledger.BlockQueue.Leaf
+import io.iohk.ethereum.ledger.BlockValidation
 import io.iohk.ethereum.mpt.MerklePatriciaTrie.MissingNodeException
 import io.iohk.ethereum.utils.BlockchainConfig
 import io.iohk.ethereum.utils.ByteStringUtils
 import io.iohk.ethereum.utils.Logger
 
-class BlockImport(
+class ConsensusImpl(
     blockchain: BlockchainImpl,
     blockchainReader: BlockchainReader,
     blockchainWriter: BlockchainWriter,
     blockQueue: BlockQueue,
     blockValidation: BlockValidation,
-    private[ledger] val blockExecution: BlockExecution,
+    blockExecution: BlockExecution,
     validationScheduler: Scheduler // Can't be implicit because of importToTop method and ambiguous of Scheduler
-) extends Logger {
+) extends Consensus
+    with Logger {
 
   /** Tries to import the block as the new best block in the chain or enqueue it for later processing.
     *
-    * @param block                 block to be imported
-    * @param blockExecutionContext threadPool on which the execution should be run
+    * @param block                   block to be imported
+    * @param blockExecutionScheduler threadPool on which the execution should be run
+    * @param blockchainConfig        blockchain configuration
     * @return One of:
-    *         - [[io.iohk.ethereum.ledger.BlockImportedToTop]] - if the block was added as the new best block
-    *         - [[io.iohk.ethereum.ledger.BlockEnqueued]] - block is stored in the [[io.iohk.ethereum.ledger.BlockQueue]]
-    *         - [[io.iohk.ethereum.ledger.ChainReorganised]] - a better new branch was found causing chain reorganisation
-    *         - [[io.iohk.ethereum.ledger.DuplicateBlock]] - block already exists either in the main chain or in the queue
-    *         - [[io.iohk.ethereum.ledger.BlockImportFailed]] - block failed to execute (when importing to top or reorganising the chain)
+    *   - [[BlockImportedToTop]] - if the block was added as the new best block
+    *   - [[BlockEnqueued]]      - block is stored in the [[io.iohk.ethereum.ledger.BlockQueue]]
+    *   - [[ChainReorganised]]   - a better new branch was found causing chain reorganisation
+    *   - [[DuplicateBlock]]     - block already exists either in the main chain or in the queue
+    *   - [[BlockImportFailed]]  - block failed to execute (when importing to top or reorganising the chain)
     */
-  def importBlock(
+  override def evaluateBranchBlock(
```

Comment:
Why is it called `evaluate`, the term for this elsewhere is `execute`, right?

Also, the comments and result type still talk of "importing", while we should probably drop that term everywhere?

Reply:
`The term elsewhere is execute, right`? No. Where did you get that idea from? Actually me and Aurelién are struggling to give this method a proper name, because it can have several outcomes. I removed the name `import` exactly because as said in the description it `Tries to import`

Reply:
There is the `BlockExecution` class for example. So this could be considered "branch execution"?

Reply:
It is not guaranteed that there will be a block or branch execution, the block may end up by just being added to the BlockQueue